aws-rds-dbcluster/src/main/java/software/amazon/rds/dbcluster/UpdateHandler.java

```diff
@@ -119,4 +139,95 @@ private boolean shouldRemoveFromGlobalCluster(
                 StringUtils.isNullOrEmpty(desiredResourceState.getGlobalClusterIdentifier());
     }
+
+    protected ProgressEvent<ResourceModel, CallbackContext> rebootCluster(
+            final AmazonWebServicesClientProxy proxy,
+            final ProxyClient<RdsClient> proxyClient,
+            final ProgressEvent<ResourceModel, CallbackContext> progress,
+            final ResourceModel resourceModel
+    ) {
+        DBCluster cluster;
+        try {
+            cluster = fetchDBCluster(proxyClient, resourceModel);
+        } catch (Exception ex) {
+            return Commons.handleException(progress, ex, DEFAULT_DB_CLUSTER_ERROR_RULE_SET);
+        }
+
+        ProgressEvent<ResourceModel, CallbackContext> result = progress;
+
+        for (final DBClusterMember dbClusterMember: cluster.dbClusterMembers()) {
+            final ProgressEvent<ResourceModel, CallbackContext> p = proxy.initiate("rds::reboot-dbcluster-member-" + dbClusterMember.dbInstanceIdentifier(),
+                            proxyClient,
+                            progress.getResourceModel(),
+                            progress.getCallbackContext())
+                    .translateToServiceRequest(model -> Translator.rebootDbInstanceRequest(dbClusterMember.dbInstanceIdentifier()))
+                    .backoffDelay(config.getBackoff())
+                    .makeServiceCall((rebootDbInstanceRequest, proxyInvocation) -> proxyInvocation.injectCredentialsAndInvokeV2(
+                            rebootDbInstanceRequest,
+                            proxyInvocation.client()::rebootDBInstance
+                    ))
+                    .stabilize((rebootDbInstanceRequest, rebootDbInstanceResponse, proxyInvocation, model, context) ->
+                            isDBInstanceStabilized(proxyClient, dbClusterMember.dbInstanceIdentifier(), model))
+                    .handleError((createRequest, exception, client, model, callbackCtx) -> Commons.handleException(
+                            ProgressEvent.progress(model, callbackCtx),
+                            exception,
+                            DEFAULT_DB_CLUSTER_ERROR_RULE_SET
+                    ))
+                    .progress();
+
+            if (p.isFailed() || p.isInProgressCallbackDelay()) {
+                return p;
+            }
+            result = p;
+        }
+        return result;
+    }
+
+    private boolean shouldRebootCluster(final ResourceModel previousResourceState, final ResourceModel desiredResourceState, boolean isRollback) {
+        return !isRollback &&
+                !Objects.equals(previousResourceState.getEngineVersion(), desiredResourceState.getEngineVersion()) &&
+                !Objects.equals(previousResourceState.getDBInstanceParameterGroupName(), desiredResourceState.getDBInstanceParameterGroupName());
+    }
+
+    private boolean isDBInstanceStabilized(final ProxyClient<RdsClient> client, final String dbInstanceIdentifier, final ResourceModel resourceModel) {
```

Comment:
I believe we should short circuit by throwing an exception if instance in available status but  `ParameterGroup` is not applied as we do not need to keep calling till stabilization timeout. There is nothing going to change if instance returned to available status but with ParameterGroup not applied. Same applies on any attributes we want to check.

Reply:
Not waiting on application of PG/CPG might trigger a stack rollback if the instance was also modified. I believe the safest thing to do is to wait for DBInstance availability. For full reasons why I am available for a verbal discussion, I think both you and Oleg might be interested.